### ISSUE TITLE
TFLite model for LiteRT GPU benchmark

### ISSUE THREAD

**Issue by nxprxf** [google-ai-edge/LiteRT#2448]
From link (https://ai.google.dev/edge/litert/next/gpu?_gl=1*ag9yln*_up*MQ..*_ga*ODk0MTcxMzI2LjE3NDkyMjIyNzk.*_ga_P1DBVKWT6V*czE3NDkyMjIyNzkkbzEkZzAkdDE3NDkyMjIyNzkkajYwJGwwJGgyNzM2Mzg3MA..) in LiteRT document, it mentioned below model benchmark list. May I know how do you convert those models to TFLite? Are you using ai-edge-torch tool (https://github.com/google-ai-edge/ai-edge-torch) to convert those models? If yes, do you have scripts to convert those models by using ai-edge-torch? 

<html>
<body>

Model | LiteRT GPU Acceleration | LiteRT GPU (ms)
-- | -- | --
hf_mms_300m | Fully delegated | 19.6
hf_mobilevit_small | Fully delegated | 8.7
hf_mobilevit_small_e2e | Fully delegated | 8.0
hf_wav2vec2_base_960h | Fully delegated | 9.1
hf_wav2vec2_base_960h_dynamic | Fully delegated | 9.8
isnet | Fully delegated | 43.1
timm_efficientnet | Fully delegated | 3.7
timm_nfnet | Fully delegated | 9.7
timm_regnety_120 | Fully delegated | 12.1
torchaudio_deepspeech | Fully delegated | 4.6
torchaudio_wav2letter | Fully delegated | 4.8
torchvision_alexnet | Fully delegated | 3.3
torchvision_deeplabv3_mobilenet_v3_large | Fully delegated | 5.7
torchvision_deeplabv3_resnet101 | Fully delegated | 35.1
torchvision_deeplabv3_resnet50 | Fully delegated | 24.5
torchvision_densenet121 | Fully delegated | 13.9
torchvision_efficientnet_b0 | Fully delegated | 3.6
torchvision_efficientnet_b1 | Fully delegated | 4.7
torchvision_efficientnet_b2 | Fully delegated | 5.0
torchvision_efficientnet_b3 | Fully delegated | 6.1
torchvision_efficientnet_b4 | Fully delegated | 7.6
torchvision_efficientnet_b5 | Fully delegated | 8.6
torchvision_efficientnet_b6 | Fully delegated | 11.2
torchvision_efficientnet_b7 | Fully delegated | 14.7
torchvision_fcn_resnet50 | Fully delegated | 19.9
torchvision_googlenet | Fully delegated | 3.9
torchvision_inception_v3 | Fully delegated | 8.6
torchvision_lraspp_mobilenet_v3_large | Fully delegated | 3.3
torchvision_mnasnet0_5 | Fully delegated | 2.4
torchvision_mobilenet_v2 | Fully delegated | 2.8
torchvision_mobilenet_v3_large | Fully delegated | 2.8
torchvision_mobilenet_v3_small | Fully delegated | 2.3
torchvision_resnet152 | Fully delegated | 15.0
torchvision_resnet18 | Fully delegated | 4.3
torchvision_resnet50 | Fully delegated | 6.9
torchvision_squeezenet1_0 | Fully delegated | 2.9
torchvision_squeezenet1_1 | Fully delegated | 2.5
torchvision_vgg16 | Fully delegated | 13.4
torchvision_wide_resnet101_2 | Fully delegated | 25.0
torchvision_wide_resnet50_2 | Fully delegated | 13.4
u2net_full | Fully delegated | 98.3
u2net_lite | Fully delegated | 51.4
hf_distil_whisper_small_no_cache | Partially delegated | 251.9
hf_distilbert | Partially delegated | 13.7
hf_tinyroberta_squad2 | Partially delegated | 17.1
hf_tinyroberta_squad2_dynamic_batch | Partially delegated | 52.1
snapml_StyleTransferNet | Partially delegated | 40.9
timm_efficientformer_l1 | Partially delegated | 17.6
timm_efficientformerv2_s0 | Partially delegated | 16.1
timm_pvt_v2_b1 | Partially delegated | 73.5
timm_pvt_v2_b3 | Partially delegated | 246.7
timm_resnest14d | Partially delegated | 88.9
torchaudio_conformer | Partially delegated | 21.5
torchvision_convnext_tiny | Partially delegated | 8.2
torchvision_maxvit_t | Partially delegated | 194.0
torchvision_shufflenet_v2 | Partially delegated | 9.5
torchvision_swin_tiny | Partially delegated | 164.4
torchvision_video_resnet2plus1d_18 | Partially delegated | 6832.0
torchvision_video_swin3d_tiny | Partially delegated | 2617.8
yolox_tiny | Partially delegated | 11.2

</body>
</html>Model	LiteRT GPU Acceleration	LiteRT GPU (ms)
[hf_mms_300m](https://huggingface.co/facebook/mms-300m)	Fully delegated	19.6
[hf_mobilevit_small](https://huggingface.co/apple/mobilevit-small)	Fully delegated	8.7
[hf_mobilevit_small_e2e](https://huggingface.co/apple/mobilevit-small)	Fully delegated	8.0
[hf_wav2vec2_base_960h](https://huggingface.co/facebook/wav2vec2-base-960h)	Fully delegated	9.1
[hf_wav2vec2_base_960h_dynamic](https://huggingface.co/facebook/wav2vec2-base-960h)	Fully delegated	9.8
[isnet](https://github.com/xuebinqin/DIS/blob/main/IS-Net/models/isnet.py)	Fully delegated	43.1
[timm_efficientnet](https://github.com/huggingface/pytorch-image-models/blob/edc37be1c3804a06c4c02b6377c214e5c66633b8/timm/models/efficientnet.py#L1974)	Fully delegated	3.7
[timm_nfnet](https://github.com/huggingface/pytorch-image-models/blob/edc37be1c3804a06c4c02b6377c214e5c66633b8/timm/models/nfnet.py#L528)	Fully delegated	9.7
[timm_regnety_120](https://github.com/huggingface/pytorch-image-models/blob/edc37be1c3804a06c4c02b6377c214e5c66633b8/timm/models/regnet.py#L1110)	Fully delegated	12.1
[torchaudio_deepspeech](https://pytorch.org/audio/stable/_modules/torchaudio/models/deepspeech.html#DeepSpeech)	Fully delegated	4.6
[torchaudio_wav2letter](https://pytorch.org/audio/stable/generated/torchaudio.models.Wav2Letter.html#torchaudio.models.Wav2Letter)	Fully delegated	4.8
[torchvision_alexnet](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.alexnet.html)	Fully delegated	3.3
[torchvision_deeplabv3_mobilenet_v3_large](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.segmentation.deeplabv3_mobilenet_v3_large.html)	Fully delegated	5.7
[torchvision_deeplabv3_resnet101](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.segmentation.deeplabv3_resnet101.html)	Fully delegated	35.1
[torchvision_deeplabv3_resnet50](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.segmentation.deeplabv3_resnet50.html)	Fully delegated	24.5
[torchvision_densenet121](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.densenet121.html)	Fully delegated	13.9
[torchvision_efficientnet_b0](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.efficientnet_b0.html)	Fully delegated	3.6
[torchvision_efficientnet_b1](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.efficientnet_b1.html)	Fully delegated	4.7
[torchvision_efficientnet_b2](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.efficientnet_b2.html)	Fully delegated	5.0
[torchvision_efficientnet_b3](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.efficientnet_b3.html)	Fully delegated	6.1
[torchvision_efficientnet_b4](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.efficientnet_b4.html)	Fully delegated	7.6
[torchvision_efficientnet_b5](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.efficientnet_b5.html)	Fully delegated	8.6
[torchvision_efficientnet_b6](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.efficientnet_b6.html)	Fully delegated	11.2
[torchvision_efficientnet_b7](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.efficientnet_b7.html)	Fully delegated	14.7
[torchvision_fcn_resnet50](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.segmentation.fcn_resnet50.html)	Fully delegated	19.9
[torchvision_googlenet](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.googlenet.html)	Fully delegated	3.9
[torchvision_inception_v3](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.inception_v3.html)	Fully delegated	8.6
[torchvision_lraspp_mobilenet_v3_large](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.segmentation.lraspp_mobilenet_v3_large.html)	Fully delegated	3.3
[torchvision_mnasnet0_5](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.mnasnet0_5.html)	Fully delegated	2.4
[torchvision_mobilenet_v2](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.mobilenet_v2.html)	Fully delegated	2.8
[torchvision_mobilenet_v3_large](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.mobilenet_v3_large.html)	Fully delegated	2.8
[torchvision_mobilenet_v3_small](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.mobilenet_v3_small.html)	Fully delegated	2.3
[torchvision_resnet152](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.resnet152.html)	Fully delegated	15.0
[torchvision_resnet18](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.resnet18.html)	Fully delegated	4.3
[torchvision_resnet50](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.resnet50.html)	Fully delegated	6.9
[torchvision_squeezenet1_0](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.squeezenet1_0.html)	Fully delegated	2.9
[torchvision_squeezenet1_1](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.squeezenet1_1.html)	Fully delegated	2.5
[torchvision_vgg16](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.vgg16.html)	Fully delegated	13.4
[torchvision_wide_resnet101_2](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.wide_resnet101_2.html#torchvision.models.wide_resnet101_2)	Fully delegated	25.0
[torchvision_wide_resnet50_2](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.wide_resnet50_2.html)	Fully delegated	13.4
[u2net_full](https://github.com/xuebinqin/U-2-Net)	Fully delegated	98.3
[u2net_lite](https://github.com/xuebinqin/U-2-Net)	Fully delegated	51.4
[hf_distil_whisper_small_no_cache](https://github.com/huggingface/distil-whisper)	Partially delegated	251.9
[hf_distilbert](https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english)	Partially delegated	13.7
[hf_tinyroberta_squad2](https://huggingface.co/deepset/tinyroberta-squad2)	Partially delegated	17.1
[hf_tinyroberta_squad2_dynamic_batch](https://huggingface.co/deepset/tinyroberta-squad2)	Partially delegated	52.1
[snapml_StyleTransferNet](https://github.com/Snapchat/snapml-templates/blob/main/Style%20Transfer/style_transfer.ipynb)	Partially delegated	40.9
[timm_efficientformer_l1](https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/efficientformer.py)	Partially delegated	17.6
[timm_efficientformerv2_s0](https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/efficientformer_v2.py)	Partially delegated	16.1
[timm_pvt_v2_b1](https://github.com/huggingface/pytorch-image-models/blob/edc37be1c3804a06c4c02b6377c214e5c66633b8/timm/models/pvt_v2.py#L525)	Partially delegated	73.5
[timm_pvt_v2_b3](https://github.com/huggingface/pytorch-image-models/blob/edc37be1c3804a06c4c02b6377c214e5c66633b8/timm/models/pvt_v2.py#L525)	Partially delegated	246.7
[timm_resnest14d](https://github.com/huggingface/pytorch-image-models/blob/edc37be1c3804a06c4c02b6377c214e5c66633b8/timm/models/resnest.py#L163)	Partially delegated	88.9
[torchaudio_conformer](https://pytorch.org/audio/stable/generated/torchaudio.models.Conformer.html)	Partially delegated	21.5
[torchvision_convnext_tiny](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.convnext_tiny.html)	Partially delegated	8.2
[torchvision_maxvit_t](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.maxvit_t.html#torchvision.models.maxvit_t)	Partially delegated	194.0
[torchvision_shufflenet_v2](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.shufflenet_v2_x1_0.html)	Partially delegated	9.5
[torchvision_swin_tiny](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.swin_t.html)	Partially delegated	164.4
[torchvision_video_resnet2plus1d_18](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.video.r2plus1d_18.html)	Partially delegated	6832.0
[torchvision_video_swin3d_tiny](https://docs.pytorch.org/vision/main/models/generated/torchvision.models.video.swin3d_t.html)	Partially delegated	2617.8
[yolox_tiny](https://github.com/Megvii-BaseDetection/YOLOX/blob/main/yolox/exp/yolox_base.py)	Partially delegated	11.2

**Comment by pkgoogle** [comment]
Hi @nxprxf, more or less yes, please review the instructions here: https://github.com/google-ai-edge/ai-edge-torch?tab=readme-ov-file#pytorch-converter. You can find examples here: https://github.com/google-ai-edge/ai-edge-torch/tree/main/ai_edge_torch/generative/examples -- if you have a question about a specific one, let me know.

**Comment by nxprxf** [comment]
@pkgoogle Thank you so much for your comments. 
I will refer to below link what you shared.  
You can find examples here: https://github.com/google-ai-edge/ai-edge-torch/tree/main/ai_edge_torch/generative/examples.

BTW, those examples are mostly focus on LLM models, may I know whether you can share some ViT or even CNN model?

**Comment by pkgoogle** [comment]
I think with ViT, it might look more like the generative models because of the Transformer blocks, CNN the first link you should be able to switch out any torchvision model, all the torchvision models are more complex than CNN, so you can just make a custom CNN in PyTorch and then follow the same instructions -- I recommend you read the docs -- then give it a try (perhaps with help from Gemini) and then let us know if you get stuck. Thanks.

**Comment by nxprxf** [comment]
@pkgoogle Thank you for your suggestion. I already managed to convert the ViT and CNN models. I saw the current quantization config support full_int8_dynamic_recipe. When I used this config to quantize the model, the weight is quantized to int8, but activation is still float32, so may I know whether there are plan to support full quantization including both activation and weight are int8?

**Comment by pkgoogle** [comment]
Hmm full int8 gets weird with indexing and slicing (only 256 available indices) so there are some cases where it doesn't quite make sense... that being said some micro devices only support 8 bits... https://github.com/tensorflow/tflite-micro might be a better fit for your situation? Perhaps take a look there to see if it has what you need, and let us know.

**Comment by nxprxf** [comment]
@pkgoogle Thank you for your quick feedback. I'm just wondering why latest ai-edge-torch tool can't support full quantization like current tflite-converter which can quantize both weight and activation. Meanwhile, we are still looking for LiteRT on our devices which are powerful application processor. For Tflite-micro, it's mainly for MCU devices.

**Comment by pkgoogle** [comment]
Hi @nxprxf, please go ahead and make a feature request for that in that repo. I just wanted to make sure we have explored what was already done, thanks. Also feel free to use the tflite-converter options, to satisfy that need as well, that library actually fully supports the "old way" as well:

```py
import tensorflow as tf

# Pass TfLite Converter quantization flags to _ai_edge_converter_flags parameter.
tfl_converter_flags = {'optimizations': [tf.lite.Optimize.DEFAULT]}

tfl_drq_model = ai_edge_torch.convert(
    torch_model, sample_args, _ai_edge_converter_flags=tfl_converter_flags
)
```

**Comment by nxprxf** [comment]
@pkgoogle Thank you for your suggestion. May I know whether you have any template for feature request? Then I can follow it to raise feature request. 

For your suggestion about TFLite-converter, yes, it can work with traditional workflow, like as below:
1. Export pytorch to onnx
2. export onnx to tensorflow save model by using onnx2tf
3. quantize and convert tensorflow save model to tflite model. 
 
As tfltie-converter can support fully quantization (W8A8) for CNN model, I can use this approach to get one fully quantized model.  But if ai-edge-torch tool can support fully quantization for CNN model, that will be great. If the ai-edge-torch mainly focus on transformer architecture or LLM model, that make sense to pending this feature until you have bandwidth to implement this feature.

**Comment by pkgoogle** [comment]
We don't have a specific one for that repo but you can try w/ this if you want:

---
name: ✨ Feature Request
about: Suggest an idea for this project
title: 'feat: [A clear and concise description of the feature]'
labels: 'enhancement, feature-request'

---

**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]

**Describe the solution you'd like**
A clear and concise description of what you want to happen.

**Describe alternatives you've considered**
A clear and concise description of any alternative solutions or features you've considered.

**Additional context**
Add any other context or screenshots about the feature request here.
